Feature/6 board create API

Leehunil opened the pull request:
resolved : #6 
## 📌 과제 설명 
사진 업로드 기능 구현 및 신고 게시글 등록 API 구현

## 👩‍💻 요구 사항과 구현 내용
- S3를 활용한 이미지 파일들 업로드 기능 구현
- 쓰레기통 추가 신고 게시글 등록 API 구현
- 쓰레기통 수정, 삭제 신고 게시글 등록 API 구현

## ✅ 피드백 반영사항  
- baseUrl, bucket과 같은 매직 스트링은 상수 처리
- S3 이미지 업로드 기능 공용으로 사용할 수 있도록 ImageService 객체로 따로 빼놓기
- Board에 Trashcan 관련 컬럼들 추가(latitude, longitude, roadNameAddress, detailedAddress, trashCategory)
- BoardImg 하나씩 save() -> saveAll(List<BoardImg>)

## ✅ PR 포인트 & 궁금한 점 
- 현재 Board에 trashcanId를 필드로 가지고 있는데 이거를 Trashcan 객체로 바꿔서 연관관계 매핑을 해두면 어떨지 고민이라서 다른 분들 의견이 궁금합니다.
- 추가 신고 게시글에는 Trashcan 정보가 포함되어야 해서 수정, 삭제 신고 게시글과 api를 분리했습니다.
- S3Config는 공유해도 될 것 같습니다. 또한 업로드하는 기능도 따로 빼놔서 공유해서 사용해도 될 것 같은데 이 부분도 확인 부탁드립니다

jhk01007 committed:
docs: Create PULL_REQUEST_TEMPLATE.md

PR 템플릿 업데이트

jhk01007 committed:
docs: PULL_REQUEST_TEMPLATE.md 삭제

디렉토리 경로 잘못잡아서 삭제합니다.

jhk01007 committed:
docs: PULL_REQUEST_TEMPLATE.md 생성

PR 템플릿 생성

Leehunil committed:
feat: S3를 활용한 이미지 업로드 기능 구현
- S3관련 메서드 사용을 위해 AWS 라이브러리 추가
- S3 bucket, baseUrl 정보 application.yml에 추가
- AWS 자격 증명을 위한 iam accessKey, secretKey 정보 application.yml에 추가
- S3 자격 증명을 얻기 위한 클라이언트 설정 클래스 S3Config 추가
- 파일이 존재해야 업로드가 가능하니 존재 여부를 파악하는 기능 추가
- 확장자를 분리해서 이미지 확장자인지 확인
- S3에 이미지 파일 업로드
- Board 저장할 때 같은 Transacion 사용할 예정이기 때문에 @Transactional은 붙이지 않는다.

Leehunil committed:
feat: S3를 활용한 이미지 업로드 기능 구현
- S3관련 메서드 사용을 위해 AWS 라이브러리 추가
- S3 bucket, baseUrl 정보 application.yml에 추가
- AWS 자격 증명을 위한 iam accessKey, secretKey 정보 application.yml에 추가
- S3 자격 증명을 얻기 위한 클라이언트 설정 클래스 S3Config 추가
- 파일이 존재해야 업로드가 가능하니 존재 여부를 파악하는 기능 추가
- 확장자를 분리해서 이미지 확장자인지 확인
- S3에 이미지 파일 업로드
- Board 저장할 때 같은 Transacion 사용할 예정이기 때문에 @Transactional은 붙이지 않는다.

Leehunil committed:
fix : 이미지 파일 하나에서 여러개로 변경

- application.yml ddl auto update로 변경
- 여러 이미지를 업로드하기 위해서 MultiPartFile에서 List<MultipartFile>로 변경

Leehunil committed:
Merge remote-tracking branch 'origin/feature/6-add-board' into feature/6-add-board

# Conflicts:
#	src/main/java/com/sscanner/team/board/service/BoardImgService.java
#	src/main/java/com/sscanner/team/board/service/BoardImgServiceImpl.java
#	src/main/java/com/sscanner/team/global/exception/ExceptionCode.java
#	src/main/resources/application.yml

Leehunil committed:
feat : 신고 게시글 등록 기능 구현

- 추가 신고 게시글과 수정, 삭제 신고 게시글 요청이 다르기 때문에 등록을 따로 진행함
- 추가 신고 게시글 등록은 trashcan도 같이 생성
- Board entity에서 BoardImg entity들 가져오기 위해서 양방향 연관관계 매핑

7zrv commented on src/main/java/com/sscanner/team/board/service/BoardImgServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+
+            BoardImg boardImg = BoardImg.makeBoardImg(board, boardImgUrl);
+
+            boardImgs.add(boardImgRepository.save(boardImg));
```
save -> saveAll()로 하는 방법은 어떠신가요
boradImg 객체들을 리스트로 모은다음 saveAll(List<BoardImg>) 와 같이하면 성능이 향상될 것 같습니다
예를들어 for문 안에서는 boardImg 리스트를 만들고 for문 밖에서  saveAll(List<BoardImg> 같은 형태로 저장하면되지 않을까 생각해봤습니다

Leehunil replied:
맞습니다!! saveAll() 사용하면 지금보다 훨씬 성능이 좋아질 것 같습니다!! 바꿔보겠습니다~

l2yujw commented on src/main/java/com/sscanner/team/board/service/BoardImgServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+    @Override
+    public String separateExt(String originalFileName) {
+        String ext = originalFileName.substring(originalFileName.lastIndexOf(".") + 1);
+
```
jpg, HEIC, jpeg, png, heic와 같은 확장자를 Set으로 등록해서
!set.contains(ext)  이런식으로 표현하면 괜찮을 수도 있지 않을까 생각이 들어서 의견 남겨봅니다!

Leehunil replied:
오 그것도 좋은 방법인 것 같습니다!! 지금 로컬 브랜치가 엉망이 돼서 다음 브랜치에서 반영해서 올려보겠습니다~

7zrv commented on src/main/java/com/sscanner/team/board/service/BoardImgServiceImpl.java:
```diff
@@ -0,0 +1,127 @@
+
+        uploadS3(file, fileName); // S3에 이미지 업로드
+
+        return baseUrl + "/" + fileName;
```
baseUrl을 상수로 바꿔보는건 어떨까요? baseUrl은 매직 스트링이라는 느낌을 받아서 제안드렸습니다!
파일 이름을 만드는 부분을 메서드화 하면 가독성이 더 좋아질 것 같다는 생각도 해봤습니다

Leehunil replied:
좋은 것 같습니다!! baseUrl은 매직 스트링이라서 상수로 관리하면 효율적일 것 같습니다. 따로 상수 클래스를 만들어서 관리를 하는게 좋을까요..??
파일 이름 생성 메서드도 분리하면 지금보다 가독성이 더 좋아질 것 같습니다!!

7zrv commented on src/main/java/com/sscanner/team/board/service/BoardServiceImpl.java:
```diff
@@ -0,0 +1,73 @@
+        List<BoardImg> boardImgs = boardImgService.saveBoardImg(savedAddBoard, files);
+        savedAddBoard.addBoardImgs(boardImgs);
+
+        return BoardCreateResponseDTO.from(savedAddBoard);
```
trashcan을 저장하는부분이 있는데 어떤 메서드인지 설명해주실수 있으신가요?
제가 아는 비즈니스 로직이랑 차이가 있어서 얘기 나눠보면 좋을것 같습니다!

Leehunil replied:
네 이 부분은 Board에 위도, 경도, 도로명 주소, 상세 주소 추가하는 걸로 하겠습니다~

Leehunil committed:
fix : 피드백 반영

- Board 테이블에 trashcan 정보 컬럼 추가
- BaseUrl, bucket 정보 상수 처리
- BoardImg 하나씩 save -> saveAll(List<BoardImg>)
- S3 업로드 기능 공유할 수 있게 따로 빼놓기

Leehunil committed:
Merge branch 'develop' into feature/6-add-board

l2yujw commented:
Merge해도 될 것 같습니다!

l2yujw reviewed (approved):
머지해도 될 것 같습니다
